### PR TITLE
Arm64 build fixes for Linux. -- In Progress --

### PR DESCRIPTION
**Pull request by jashook**
Modifies code that is not platform specific, please do not merge until at least all tests pass.

Includes several changes to fix the build break for arm64 on linux. Currently have it building on arm64 hardware, however, need to check that it does not regress other targets or the cross build. Also note in order to get it to build some things may have been removed that are unnecessary if that is encountered please notify me to better solutions.

!Note

This fixes the build, however, the resulting corerun will not be able to initialize the VM properly.

### CODE REVIEW

**Comment by jashook** [comment]
@mmitche  PTAL

File: src/pal/src/include/pal/virtual.h [hunk 1]
```diff
@@ -65,6 +65,8 @@ enum VIRTUAL_CONSTANTS
     /* Page manipulation constants. */
 #ifdef __sparc__
     VIRTUAL_PAGE_SIZE       = 0x2000,
+#elif __aarch64__
```
Comment (jashook) [inline src/pal/src/include/pal/virtual.h]: This causes the vm assertion. Will remove.

File: src/inc/staticcontract.h [hunk 2]
```diff
@@ -237,7 +237,11 @@ namespace StaticContract
 {
     struct ScanThrowMarkerStandard
     {
+#if !defined(PLATFORM_UNIX)
         __declspec(noinline) ScanThrowMarkerStandard()
```
Comment (janvorli) [inline src/inc/staticcontract.h]: You can use the NOINLINE macro which is defined like this:
#ifdef _MSC_VER
#define NOINLINE __declspec(noinline)
#elif defined __GNUC__
#define NOINLINE __attribute__ ((noinline))
#else
#define NOINLINE
#endif

File: src/pal/src/exception/seh.cpp [hunk 3]
```diff
@@ -259,10 +259,10 @@ bool CatchHardwareExceptionHolder::IsEnabled()
 --*/
 
 #ifdef __llvm__
-__thread 
-#else // __llvm__
+__thread
+#else
```
Comment (janvorli) [inline src/pal/src/exception/seh.cpp]: Was the comment on the `#else` removed by accident?

File: src/pal/src/exception/signal.cpp [hunk 4]
```diff
@@ -441,7 +441,7 @@ static void inject_activation_handler(int code, siginfo_t *siginfo, void *contex
             {
                 g_activationFunction(&winContext);
             }
-
+
```
Comment (janvorli) [inline src/pal/src/exception/signal.cpp]: Unintentional blanks?

File: src/pal/src/objmgr/palobjbase.hpp [hunk 5]
```diff
@@ -185,6 +185,9 @@ namespace CorUnix
         ReleaseReference(
             CPalThread *pthr
             );
+
```
Comment (janvorli) [inline src/pal/src/objmgr/palobjbase.hpp]: Unintentional blank lines?

File: src/pal/tools/gen-buildsys-clang.sh [hunk 6]
```diff
@@ -131,10 +160,14 @@ if [[ -n "$CROSSCOMPILE" ]]; then
     cmake_extra_defines="$cmake_extra_defines -DCMAKE_TOOLCHAIN_FILE=$1/cross/$build_arch/toolchain.cmake"
 fi
 
+echo "cmake -DCMAKE_USER_MAKE_RULES_OVERRIDE=$1/src/pal/tools/clang-compiler-override.txt -DCMAKE_AR=$llvm_ar -DCMAKE_C_COMPILER=$CC -DCMAKE_CXX_COMPILER=$CXX -DCMAKE_LINKER=$llvm_link -DCMAKE_NM=$llvm_nm -DCMAKE-OBJDUMP=$llvm_objdump -DCMAKE_BUILD_TYPE=$buildtype -DCMAKE_ENABLE_CODE_COVERAGE=$code_coverage $cmake_extra_defines $1"
```
Comment (janvorli) [inline src/pal/tools/gen-buildsys-clang.sh]: I guess this will go away when you're ready to merge in, right?

File: src/utilcode/clrhost_nodependencies.cpp [hunk 7]
```diff
@@ -383,7 +383,7 @@ FastFreeInProcessHeapFunc __ClrFreeInProcessHeap = (FastFreeInProcessHeapFunc) C
 
 const NoThrow nothrow = { 0 };
 
-#ifdef __llvm__
+#if defined(__llvm__) && !defined(__aarch64__)
```
Comment (janvorli) [inline src/utilcode/clrhost_nodependencies.cpp]: Why is this needed for ARM64? It worries me since that would expose the overriden new / delete to the outside world and it should be internal only.

File: src/vm/threads.cpp [hunk 8]
```diff
@@ -303,8 +303,8 @@ bool Thread::DetectHandleILStubsForDebugger()
 extern "C" {
 #ifndef __llvm__
 __declspec(thread)
-#else // !__llvm__
-__thread 
+#else // __llvm__
```
Comment (janvorli) [inline src/vm/threads.cpp]: The comment should have the `!` in it. The convention is to put the same comment on `#if`, `#else` and `#end`

File: src/vm/threads.h [hunk 9]
```diff
@@ -3740,13 +3739,15 @@ class Thread: public IUnknown
     // to identifying redirection targets
     static BOOL IsAddrOfRedirectFunc(void * pFuncAddr);
 
+#endif //defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
```
Comment (janvorli) [inline src/vm/threads.h]: The convention is to have just FEATURE_HIJACK && !PLATFORM_UNIX here.

File: src/vm/threads.h [hunk 10]
```diff
@@ -2878,9 +2878,7 @@ class Thread: public IUnknown
         STR_SwitchedOut,
     };
 
-#if defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
```
Comment (janvorli) [inline src/vm/threads.h]: Why was this `#ifdef` removed? The method is only for FEATURE_HIJACK && PLATFORM_UNIX. The definition of the method is under the same `#ifdef`.

File: src/vm/threadsuspend.cpp [hunk 11]
```diff
@@ -8342,7 +8342,7 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
 #endif //TIME_SUSPEND
 }
 
-#if defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
+#if !defined(_TARGET_ARM64) && defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
```
Comment (jkotas) [inline src/vm/threadsuspend.cpp]: This should not ifdefed out for TARGET_ARM64. If there are problems in this code that you do not want to fix right now, disable FEATURE_HIJACK in the central CMakeFile and add TODO next to it.

File: src/vm/threads.h (hunk 9 above)
Comment (jkotas) [inline src/vm/threads.h]: And also, the #endif should not be moved from its original location - because of it matched how the implementation is ifdefed out.

File: src/vm/threads.h [hunk 12]
```diff
@@ -2878,9 +2878,7 @@ class Thread: public IUnknown
...
     bool InjectGcSuspension();
```
Comment (jkotas) [inline src/vm/threads.h]: Why remove this ifdef?

File: src/vm/threadsuspend.cpp (hunk 11 above)
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: I don't think we should exclude it for ARM64, we need to implement activation injection for it too.

File: src/vm/threadsuspend.cpp [hunk 13]
```diff
@@ -8487,12 +8487,12 @@ bool Thread::InjectGcSuspension()
...
 }
 
-#endif // FEATURE_HIJACK && PLATFORM_UNIX
+#endif // !_TARGET_ARM64 && FEATURE_HIJACK && PLATFORM_UNIX
 
 // Initialize thread suspension support
 void ThreadSuspend::Initialize()
 {
-#if defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
+#if defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX) && defined(_TARGET_AMD64_)
```
Comment (janvorli) [inline src/vm/threadsuspend.cpp]: We don't need to add the AMD64 ifdef here. The injection should be implemented for all non-windows systems.

File: src/vm/arm64/stubs.cpp [hunk 14]
```diff
@@ -798,6 +798,24 @@ void ResumableFrame::UpdateRegDisplay(const PREGDISPLAY pRD)
     pRD->ControlPC = m_Regs->Pc;
     pRD->SP = m_Regs->Sp;
 
+    pRD->volatileCurrContextPointers.X[0] = &m_Regs->X0;
+    pRD->volatileCurrContextPointers.X[1] = &m_Regs->X1;
```
Comment (jkotas) [inline src/vm/arm64/stubs.cpp]: You should update the definition of the CONTEXT structure in the PAL to what is in Windows headers instead, and undo this change.

File: src/pal/src/exception/seh.cpp (hunk 3 above)
Comment (jkotas) [inline src/pal/src/exception/seh.cpp]: BTW: The __llvm__  ifdef is not needed here because of PAL is always building with LLVM. You can just remove it.

File: src/pal/inc/rt/safecrt.h [hunk 15]
```diff
@@ -136,7 +136,7 @@ typedef _W64 unsigned int   uintptr_t;
 #define _UINTPTR_T_DEFINED
 #endif
 
-#if defined(_MSC_VER) || (defined(__GNUC__) && (__GNUC__ >= 3))
+#if defined(_MSC_VER) || (defined(__GNUC__) && (__GNUC__ >= 3) && !(defined(PLATFORM_UNIX) && defined(__aarch64__)))
 #define SAFECRT_DEPRECATED __declspec(deprecated)
```
Comment (jkotas) [inline src/pal/inc/rt/safecrt.h]: It would be better to delete SAFECRT_DEPRECATED and errcode that it is used on instead.

File: tests/src/Common/Platform/platformdefines.cpp [hunk 16]
```diff
@@ -331,10 +331,10 @@ void TP_DebugBreak()
 {
 #ifdef _PPC_
 	__asm__ __volatile__("trap");
-#elif defined(_ARM_)
-    __asm__ __volatile__("bkpt");
-#else
+#elif defined(__amd64__) || defined(_M_AMD64) || defined(__i386) || defined(_M_IX86)
 	__asm__ __volatile__("int $3");
+#elif defined(__arm__)
+    __asm__ __volatile__("bkpt #0");
```
Comment (benpye) [inline tests/src/Common/Platform/platformdefines.cpp]: If we want arm to use `bkpt #0` then you may want to change https://github.com/dotnet/coreclr/blob/master/src/pal/inc/unixasmmacrosarm.inc#L192

File: build.sh [hunk 17]
```diff
@@ -261,6 +263,10 @@ for i in "$@"
         arm64)
         __BuildArch=arm64
         __MSBuildBuildArch=arm64
+        
+        # if we are on arm64 then we need clang 3.8
+
+        __ClangMinorVersion=8
```
Comment (kangaroo) [inline build.sh]: Why is this needed?  clang-3.5 works fine on my board

File: src/utilcode/clrhost_nodependencies.cpp (hunk 7 above)
Comment (kangaroo) [inline src/utilcode/clrhost_nodependencies.cpp]: None of this is needed on my board.  Is this a clang-3.8 ism?

**Comment by kangaroo** [comment]
The majority of these changes appear related to the bump to clang-3.8 to me (which I don't know the reason for).  That said, I'd prefer to split this into two patches if we want to make the code clang 3.8 safe.

1 patch for clang-3.8 changes
1 patch for architecture specific changes

File: build.sh (hunk 17 above)
Comment (jkotas) [inline build.sh]: +1 

clang 3.8 is moving target because of it was not released yet. It is not a good idea to require it unless it is absolutely needed.

File: tests/src/Common/Platform/platformdefines.cpp (hunk 16 above)
Comment (jkotas) [inline tests/src/Common/Platform/platformdefines.cpp]: It would be better to just delete this method. It is not used anywhere.

Comment (benpye) [inline tests/src/Common/Platform/platformdefines.cpp]: I just wondered if there was some motivation behind the change, it may apply to the assembly macro also.

File: src/vm/threadsuspend.cpp (hunk 11 above)
Comment (kangaroo) [inline src/vm/threadsuspend.cpp]: In addition, I think th define is incorrect (note the missing trailing _ or incorrect leading _)

**Comment by jashook** [comment]
The dependency on 3.8 was chosen after 3.5 failed compiling due an internal compiler error. On my machine aarch64 fedora 22 llvm-3.5 fails instantly.

File: build.sh (hunk 17 above)
Comment (jashook) [inline build.sh]: See last comment

**Comment by mmitche** [comment]
@dotnet-bot test arm

**Comment by mmitche** [comment]
@jashook when you have the feedback resolved comment that phrase again.

**Comment by kangaroo** [comment]
@jashook clang-3.5 is working just fine on a recent linaro (ubuntu/debian based) distro here.  Perhaps something is wrong with your Fedora image?  Try ubuntu?  Again I'd like to request that all the clang-isms are split into a separate patch from the architecture changes.  The clang-isms will have trickle down effects to all platforms.

**Comment by jashook** [comment]
I will investigate that further then.

File: build.sh (hunk 17 above)
Comment (jkotas) [inline build.sh]: Can we pick 3.7 instead - because of it is a released version?
